File: ios/OfflineWorkoutTracker/Bridge/OwtBridge.swift
import Foundation
import OwtBridgeC

final class OwtBridge {
    static let shared = OwtBridge()
    private init() {}

    func initialize() {
        let supportDir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        try? FileManager.default.createDirectory(at: supportDir, withIntermediateDirectories: true)
        let dbPath = supportDir.appendingPathComponent("owt.db").path
        owt_init(dbPath)
    }

    func close() {
        owt_close()
    }

    // MARK: - Exercises

    func addExercise(name: String, category: String, muscleGroup: String, notes: String) -> Int64 {
        return owt_add_exercise(name, category, muscleGroup, notes)
    }

    func listExercises(filter: String = "") -> [Exercise] {
        var buffer = [OwtExercise](repeating: OwtExercise(), count: 500)
        let count = owt_list_exercises(filter, &buffer, 500)
        return (0..<Int(count)).map { Exercise(from: buffer[$0]) }
    }

    func updateExercise(id: Int64, name: String, category: String, muscleGroup: String, notes: String) {
        owt_update_exercise(id, name, category, muscleGroup, notes)
    }

    func deleteExercise(id: Int64) {
        owt_delete_exercise(id)
    }

    // MARK: - Workouts

    func startWorkout(name: String = "") -> Int64 {
        return owt_start_workout(name)
    }

    func finishWorkout(id: Int64) {
        owt_finish_workout(id)
    }

    func getActiveWorkout() -> Workout? {
        var w = OwtWorkout()
        let found = owt_get_active_workout(&w)
        guard found != 0 else { return nil }
        return Workout(from: w)
    }

    func listWorkouts(limit: Int = 20, offset: Int = 0) -> [Workout] {
        var buffer = [OwtWorkout](repeating: OwtWorkout(), count: 200)
        let count = owt_list_workouts(Int32(limit), Int32(offset), &buffer, 200)
        return (0..<Int(count)).map { Workout(from: buffer[$0]) }
    }

    func deleteWorkout(id: Int64) {
        owt_delete_workout(id)
    }

    // MARK: - Sets

    func addSet(workoutId: Int64, exerciseId: Int64, order: Int, reps: Int, weight: Double, rpe: Double, durationSecs: Int = 0, restSecs: Int = 0) -> Int64 {
        return owt_add_set(workoutId, exerciseId, Int32(order), Int32(reps), weight, rpe, Int32(durationSecs), Int32(restSecs))
    }

    func getSetsForWorkout(workoutId: Int64) -> [WorkoutSet] {
        var buffer = [OwtWorkoutSet](repeating: OwtWorkoutSet(), count: 500)
        let count = owt_get_sets_for_workout(workoutId, &buffer, 500)
        return (0..<Int(count)).map { WorkoutSet(from: buffer[$0]) }
    }

    func updateSet(id: Int64, reps: Int, weight: Double, rpe: Double, durationSecs: Int = 0, restSecs: Int = 0) {
        owt_update_set(id, Int32(reps), weight, rpe, Int32(durationSecs), Int32(restSecs))
    }

    func deleteSet(id: Int64) {
        owt_delete_set(id)
    }

    // MARK: - Stats

    func getStats(exerciseId: Int64) -> ExerciseStats? {
        var s = OwtExerciseStats()
        let found = owt_get_stats(exerciseId, &s)
        guard found != 0 else { return nil }
        return ExerciseStats(from: s)
    }

    func getProgression(exerciseId: Int64, sessions: Int = 20) -> [ProgressionPoint] {
        var buffer = [OwtProgressionPoint](repeating: OwtProgressionPoint(), count: 100)
        let count = owt_get_progression(exerciseId, Int32(sessions), &buffer, 100)
        return (0..<Int(count)).map { ProgressionPoint(from: buffer[$0]) }
    }

    // MARK: - Templates

    func createTemplate(name: String, notes: String = "") -> Int64 {
        return owt_create_template(name, notes)
    }

    func listTemplates() -> [WorkoutTemplate] {
        var buffer = [OwtWorkoutTemplate](repeating: OwtWorkoutTemplate(), count: 100)
        let count = owt_list_templates(&buffer, 100)
        return (0..<Int(count)).map { WorkoutTemplate(from: buffer[$0]) }
    }

    func getTemplateSets(templateId: Int64) -> [TemplateSet] {
        var buffer = [OwtTemplateSet](repeating: OwtTemplateSet(), count: 200)
        let count = owt_get_template_sets(templateId, &buffer, 200)
        return (0..<Int(count)).map { TemplateSet(from: buffer[$0]) }
    }

    func deleteTemplate(id: Int64) {
        owt_delete_template(id)
    }

    func addTemplateSet(templateId: Int64, exerciseId: Int64, order: Int, reps: Int, weight: Double, rpe: Double, durationSecs: Int = 0, restSecs: Int = 0) -> Int64 {
        return owt_add_template_set(templateId, exerciseId, Int32(order), Int32(reps), weight, rpe, Int32(durationSecs), Int32(restSecs))
    }

    func deleteTemplateSet(id: Int64) {
        owt_delete_template_set(id)
    }

    func swapTemplateSetOrder(idA: Int64, orderA: Int, idB: Int64, orderB: Int) {
        owt_swap_template_set_order(idA, Int32(orderA), idB, Int32(orderB))
    }

    func startWorkoutFromTemplate(templateId: Int64, name: String = "") -> Int64 {
        return owt_start_workout_from_template(templateId, name)
    }

    // MARK: - Weight Conversion

    func toDisplayWeight(_ storedKg: Double) -> Double {
        return owt_to_display_weight(storedKg, getWeightUnit())
    }

    func toStorageWeight(_ displayValue: Double) -> Double {
        return owt_to_storage_weight(displayValue, getWeightUnit())
    }

    // MARK: - Settings

    func getWeightUnit() -> String {
        return String(cString: owt_get_weight_unit())
    }

    func setWeightUnit(_ unit: String) {
        owt_set_weight_unit(unit)
    }

    func getOneRepMax(exerciseId: Int64) -> Double {
        return owt_get_one_rep_max(exerciseId)
    }

    func setOneRepMax(exerciseId: Int64, weight: Double) {
        owt_set_one_rep_max(exerciseId, weight)
    }

    func isSetupComplete() -> Bool {
        return owt_is_setup_complete() != 0
    }

    func setSetupComplete(_ complete: Bool) {
        owt_set_setup_complete(complete ? 1 : 0)
    }

    // MARK: - Defaults

    func seedDefaultExercises() {
        owt_seed_default_exercises()
    }

    func seedDefaultTemplates() {
        owt_seed_default_templates()
    }

    // MARK: - Export/Import

    func exportToJson() -> String {
        return String(cString: owt_export_to_json())
    }

    func previewImport(json: String) -> ImportSummary? {
        var s = OwtImportSummary()
        let found = owt_preview_import(json, &s)
        guard found != 0 else { return nil }
        return ImportSummary(from: s)
    }

    func importFromJson(json: String) -> String {
        return String(cString: owt_import_from_json(json))
    }
}

File: android/app/src/main/java/com/timsippell/owt/bridge/OwtBridge.kt
package com.timsippell.owt.bridge

import android.content.Context

object OwtBridge {
    init {
        System.loadLibrary("owt-jni")
    }

    data class Exercise(
        val id: Long,
        val name: String,
        val category: String,
        val muscleGroup: String,
        val notes: String
    )

    data class Workout(
        val id: Long,
        val name: String,
        val startedAt: String,
        val finishedAt: String,
        val notes: String,
        val setCount: Int
    )

    data class WorkoutSet(
        val id: Long,
        val workoutId: Long,
        val exerciseId: Long,
        val order: Int,
        val reps: Int,
        val weight: Double,
        val rpe: Double,
        val durationSecs: Int,
        val restSecs: Int
    )

    data class ExerciseStats(
        val exerciseId: Long,
        val estimated1rm: Double,
        val bestWeight: Double,
        val totalVolume: Double,
        val sessionCount: Int
    )

    data class WorkoutTemplate(
        val id: Long,
        val name: String,
        val notes: String,
        val setCount: Int
    )

    data class TemplateSet(
        val id: Long,
        val templateId: Long,
        val exerciseId: Long,
        val order: Int,
        val reps: Int,
        val weight: Double,
        val rpe: Double,
        val durationSecs: Int,
        val restSecs: Int
    )

    data class ProgressionPoint(
        val date: String,
        val estimated1rm: Double,
        val bestSetWeight: Double,
        val sessionVolume: Double
    )

    data class ImportSummary(
        val newExercises: Int,
        val existingExercises: Int,
        val workouts: Int,
        val workoutSets: Int,
        val templates: Int,
        val templateSets: Int
    )

    fun init(context: Context) {
        val dbPath = context.getDatabasePath("owt.db").absolutePath
        context.getDatabasePath("owt.db").parentFile?.mkdirs()
        nativeInit(dbPath)
    }

    fun close() = nativeClose()

    fun addExercise(name: String, category: String, muscleGroup: String, notes: String): Long =
        nativeAddExercise(name, category, muscleGroup, notes)

    fun listExercises(filter: String = ""): List<Exercise> =
        nativeListExercises(filter)?.toList() ?: emptyList()

    fun updateExercise(id: Long, name: String, category: String, muscleGroup: String, notes: String) =
        nativeUpdateExercise(id, name, category, muscleGroup, notes)

    fun deleteExercise(id: Long) = nativeDeleteExercise(id)

    fun startWorkout(name: String = ""): Long = nativeStartWorkout(name)
    fun finishWorkout(id: Long) = nativeFinishWorkout(id)
    fun getActiveWorkout(): Workout? = nativeGetActiveWorkout()
    fun listWorkouts(limit: Int = 20, offset: Int = 0): List<Workout> =
        nativeListWorkouts(limit, offset)?.toList() ?: emptyList()
    fun deleteWorkout(id: Long) = nativeDeleteWorkout(id)

    fun addSet(workoutId: Long, exerciseId: Long, order: Int, reps: Int, weight: Double, rpe: Double, durationSecs: Int = 0, restSecs: Int = 0): Long =
        nativeAddSet(workoutId, exerciseId, order, reps, weight, rpe, durationSecs, restSecs)
    fun getSetsForWorkout(workoutId: Long): List<WorkoutSet> =
        nativeGetSetsForWorkout(workoutId)?.toList() ?: emptyList()
    fun deleteSet(id: Long) = nativeDeleteSet(id)

    fun updateSet(id: Long, reps: Int, weight: Double, rpe: Double, durationSecs: Int = 0, restSecs: Int = 0) =
        nativeUpdateSet(id, reps, weight, rpe, durationSecs, restSecs)

    fun createTemplate(name: String, notes: String = ""): Long =
        nativeCreateTemplate(name, notes)
    fun listTemplates(): List<WorkoutTemplate> =
        nativeListTemplates()?.toList() ?: emptyList()
    fun getTemplateSets(templateId: Long): List<TemplateSet> =
        nativeGetTemplateSets(templateId)?.toList() ?: emptyList()
    fun updateTemplate(id: Long, name: String, notes: String = "") =
        nativeUpdateTemplate(id, name, notes)
    fun deleteTemplate(id: Long) = nativeDeleteTemplate(id)
    fun addTemplateSet(templateId: Long, exerciseId: Long, order: Int, reps: Int, weight: Double, rpe: Double, durationSecs: Int = 0, restSecs: Int = 0): Long =
        nativeAddTemplateSet(templateId, exerciseId, order, reps, weight, rpe, durationSecs, restSecs)
    fun updateTemplateSet(id: Long, reps: Int, weight: Double, rpe: Double, durationSecs: Int = 0, restSecs: Int = 0) =
        nativeUpdateTemplateSet(id, reps, weight, rpe, durationSecs, restSecs)
    fun deleteTemplateSet(id: Long) = nativeDeleteTemplateSet(id)
    fun swapTemplateSetOrder(idA: Long, orderA: Int, idB: Long, orderB: Int) =
        nativeSwapTemplateSetOrder(idA, orderA, idB, orderB)
    fun startWorkoutFromTemplate(templateId: Long, name: String = ""): Long =
        nativeStartWorkoutFromTemplate(templateId, name)

    fun getStats(exerciseId: Long): ExerciseStats? = nativeGetStats(exerciseId)
    fun getProgression(exerciseId: Long, sessions: Int = 20): List<ProgressionPoint> =
        nativeGetProgression(exerciseId, sessions)?.toList() ?: emptyList()

    // Weight conversion
    fun toDisplayWeight(storedKg: Double, unit: String): Double = nativeToDisplayWeight(storedKg, unit)
    fun toStorageWeight(displayValue: Double, unit: String): Double = nativeToStorageWeight(displayValue, unit)

    // Settings
    fun getWeightUnit(): String = nativeGetWeightUnit()
    fun setWeightUnit(unit: String) = nativeSetWeightUnit(unit)
    fun getOneRepMax(exerciseId: Long): Double = nativeGetOneRepMax(exerciseId)
    fun setOneRepMax(exerciseId: Long, weight: Double) = nativeSetOneRepMax(exerciseId, weight)
    fun isSetupComplete(): Boolean = nativeIsSetupComplete()
    fun setSetupComplete(complete: Boolean) = nativeSetSetupComplete(complete)

    // Defaults
    fun seedDefaultExercises() = nativeSeedDefaultExercises()
    fun seedDefaultTemplates() = nativeSeedDefaultTemplates()

    // Export/Import
    fun exportToJson(): String = nativeExportToJson()
    fun previewImport(json: String): ImportSummary? = nativePreviewImport(json)
    fun importFromJson(json: String): String = nativeImportFromJson(json)

    private external fun nativeInit(dbPath: String)
    private external fun nativeClose()
    private external fun nativeAddExercise(name: String, category: String, muscleGroup: String, notes: String): Long
    private external fun nativeListExercises(filter: String): Array<Exercise>?
    private external fun nativeUpdateExercise(id: Long, name: String, category: String, muscleGroup: String, notes: String)
    private external fun nativeDeleteExercise(id: Long)
    private external fun nativeStartWorkout(name: String): Long
    private external fun nativeFinishWorkout(id: Long)
    private external fun nativeGetActiveWorkout(): Workout?
    private external fun nativeListWorkouts(limit: Int, offset: Int): Array<Workout>?
    private external fun nativeDeleteWorkout(id: Long)
    private external fun nativeAddSet(workoutId: Long, exerciseId: Long, order: Int, reps: Int, weight: Double, rpe: Double, durationSecs: Int, restSecs: Int): Long
    private external fun nativeGetSetsForWorkout(workoutId: Long): Array<WorkoutSet>?
    private external fun nativeDeleteSet(id: Long)
    private external fun nativeUpdateSet(id: Long, reps: Int, weight: Double, rpe: Double, durationSecs: Int, restSecs: Int)
    private external fun nativeCreateTemplate(name: String, notes: String): Long
    private external fun nativeListTemplates(): Array<WorkoutTemplate>?
    private external fun nativeGetTemplateSets(templateId: Long): Array<TemplateSet>?
    private external fun nativeUpdateTemplate(id: Long, name: String, notes: String)
    private external fun nativeDeleteTemplate(id: Long)
    private external fun nativeAddTemplateSet(templateId: Long, exerciseId: Long, order: Int, reps: Int, weight: Double, rpe: Double, durationSecs: Int, restSecs: Int): Long
    private external fun nativeUpdateTemplateSet(id: Long, reps: Int, weight: Double, rpe: Double, durationSecs: Int, restSecs: Int)
    private external fun nativeDeleteTemplateSet(id: Long)
    private external fun nativeSwapTemplateSetOrder(idA: Long, orderA: Int, idB: Long, orderB: Int)
    private external fun nativeStartWorkoutFromTemplate(templateId: Long, name: String): Long
    private external fun nativeGetStats(exerciseId: Long): ExerciseStats?
    private external fun nativeGetProgression(exerciseId: Long, sessions: Int): Array<ProgressionPoint>?
    private external fun nativeToDisplayWeight(storedKg: Double, unit: String): Double
    private external fun nativeToStorageWeight(displayValue: Double, unit: String): Double
    private external fun nativeGetWeightUnit(): String
    private external fun nativeSetWeightUnit(unit: String)
    private external fun nativeGetOneRepMax(exerciseId: Long): Double
    private external fun nativeSetOneRepMax(exerciseId: Long, weight: Double)
    private external fun nativeIsSetupComplete(): Boolean
    private external fun nativeSetSetupComplete(complete: Boolean)
    private external fun nativeSeedDefaultExercises()
    private external fun nativeSeedDefaultTemplates()
    private external fun nativeExportToJson(): String
    private external fun nativePreviewImport(json: String): ImportSummary?
    private external fun nativeImportFromJson(json: String): String
}
